client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -172,11 +179,12 @@ private void blockFor(long timeoutMs) {
     }
 
     private PositionInternal getPosition() {
-        Map<Segment, Long> ownedSegments = new HashMap<>(sealedSegments);
+        Map<Segment, Long> ownedSegments = new HashMap<>(sealedSegments.size() + readers.size());
```

Comment:
Let's make ownedSegments a List<Entry<Segment, Long>> and adapt positionImpl to store this. This will be much cheaper to construct.

Reply:
I did that change but didn't show much improvement. Instead, I opted for another kind of optimization in which `ownedSegments` is re-used and we batch the changes of offsets for every read. A set of Position objects that reference the same `ownedSegments` object, will also have a reference to a list of ` List<Entry<Segment, Long>> segmentOffsetUpdates`. Each Position object will then lazily apply the updates to segment offsets up to the point the event was read. This has proved to make readers much faster reading small events, specially when multiple segments are involved.